tests/unit/machine/all_machines_unittest.cc

```diff
@@ -0,0 +1,384 @@
+		break;
+	case PT_REGRESSION:
+		ec = new CMeanSquaredError();
+		break;
+	default:
+		SG_SNOTIMPLEMENTED
+		break;
+	}
+
+	auto cv = new CCrossValidation(machine, data.first, data.second, ss, ec);
+	cv->set_num_runs(3);
+	SG_REF(cv);
+
+	return cv;
+}
+
+std::pair<CDenseFeatures<float64_t>*, CDenseLabels*> generate_data(const CMachine* machine)
+{
+	auto N = 50;
+	auto D = 3;
+
+	std::mt19937_64 prng(57);
+	NormalDistribution<float64_t> randn;
+	UniformRealDistribution<float64_t> rand(0,1);
+	UniformIntDistribution<int32_t> randi(0,2);
+
+	SGMatrix<float64_t> X(D,N);
+	for (auto i : range(D*N))
+		X.matrix[i] = randn(prng);
+	auto features = new CDenseFeatures<float64_t>(X);
+
+	SGVector<float64_t> y_reg(N);
+	SGVector<float64_t> y_binary(N);
+	SGVector<float64_t> y_mc(N);
+
+	for (auto i : range(N))
+	{
+		auto redux = linalg::mean(X.get_column(i));
+
+		y_reg[i] = redux + std::sin(redux) + 1;
+		y_mc[i] = redux<0 ? 0 : 1;
+		y_binary[i] = y_mc[i] * 2 -1;
+
+		// noise
+		y_reg[i] +=  randn(prng)*0.1;
+		if (rand(prng)>0.1)
+		{
+			y_binary[i] *= (-1);
+			y_mc[i] = (int32_t(y_mc[i]) + randi(prng)) % 3;
+		}
+	}
+
+	CDenseLabels* labels;
+	auto pt = machine->get_machine_problem_type();
+	switch (pt)
+	{
+	case PT_BINARY:
+	case PT_CLASS:
+	{
+		labels = new CBinaryLabels(y_binary);
+		break;
+	}
+
+	case PT_MULTICLASS:
+	{
+		labels = new CMulticlassLabels(y_mc);
+		break;
+	}
+
+	case PT_REGRESSION:
+		labels = new CRegressionLabels(y_reg);
+		break;
+
+	default:
+		SG_SERROR("Unsupported problem type: %d\n", pt);
+	}
+
+	SG_REF(features)
+	SG_REF(labels)
+	return make_pair(features, labels);
+}
+
+void serialize_machine(CMachine* machine, std::string& fname)
+{
+	auto fs = io::FileSystemRegistry::instance();
+	std::string class_name = machine->get_name();
+	fname = "shogun-unittest-AllMachines-trained_model_seiralization_consistency-" + class_name +
+			   ".XXXXXX";
+	generate_temp_filename(const_cast<char*>(fname.c_str()));
+
+	SG_REF(machine);
+	EXPECT_FALSE(fs->file_exists(fname));
+	std::unique_ptr<io::WritableFile> file;
+	EXPECT_FALSE(fs->new_writable_file(fname, &file));
+	auto fos = some<io::CFileOutputStream>(file.get());
+	auto serializer = some<io::CBitserySerializer>();
+	serializer->attach(fos);
+	serializer->write(wrap<CSGObject>(machine));
+}
+
+CMachine* deserialize_machine(std::string fname)
+{
+	auto fs = io::FileSystemRegistry::instance();
+	std::unique_ptr<io::RandomAccessFile> raf;
+	EXPECT_FALSE(fs->new_random_access_file(fname, &raf));
+	auto fis = some<io::CFileInputStream>(raf.get());
+	auto deserializer = some<io::CBitseryDeserializer>();
+	deserializer->attach(fis);
+	auto deser_obj = deserializer->read_object();
+	bool delete_success = !fs->delete_file(fname);
+	EXPECT_TRUE(delete_success);
+
+	return dynamic_cast<CMachine*>(deser_obj.get());
+}
+
+// TODO, generate this automatically, like in trained_model_serialization
+std::set<string> all_machines = {"LibSVM", "Perceptron", "LibLinear",
```

Comment:
at least make a function that returns this set instead of having it here a global var....

Reply:
will do once this extraction is automated